ykllvmwrap/src/jitmodbuilder.cc

```diff
@@ -1470,38 +1501,72 @@ class JITModBuilder {
     for (size_t Idx = 0; Idx < InpTrace.Length(); Idx++) {
       // Update the previously executed BB in the most-recent frame (if it's
       // mappable).
-      optional<IRBlock> MaybeIB = InpTrace[Idx];
+      TraceLoc Loc = InpTrace[Idx];
 
-      // If we previously saw a call to foreign code, there will now be an
-      // "unmappable hole" in the trace.
-      if (ExpectUnmappable) {
-        assert(!MaybeIB.has_value());
-        // FIXME: Allow foreign code to call back to mappable code.
+      if (UnmappableRegion *UR = Loc.getUnmappableRegion()) {
+        // The trace entered a region of unmappable foreign code.
         //
-        // We've just seen an unmappable region in the trace. There are never
-        // two unmappable regions in a row, so the next block is guaranteed to
-        // be mappable. Since we don't yet allow foreign code to "call back"
-        // into mappable code, we assume that the subsequent mappable block is
-        // the result of the foreign code returning to mappable code.
-        assert(!CallStack.curFrame().getMappableFrame());
-        CallStack.popFrame();
-        ExpectUnmappable = false;
-        tryStopOutlining();
+        // There are never two unmappable regions in a row, so the next block
```

Comment:
> There are never two unmappable regions in a row

How do we know that?

Reply:
Via documentation and assertion :)
 
- https://github.com/ykjit/yk/blob/master/yktrace/src/hwt/mapper.rs#L139
- https://github.com/ykjit/yk/blob/5d2c6a28d7f3ee54f264e4678e12c38ab30c4a65/ykllvmwrap/src/jitmodbuilder.cc#L1370-L1374

Reply:
Ah, can we reference this (even if we just say "as noted above there are never two ...")?